solutions/usaco-1211.mdx

```diff
@@ -0,0 +1,101 @@
+---
+id: usaco-1211
+source: USACO Gold 2022 February
+title: Moo Network
+author: Jeffrey Zhang
+---
+
+[Official Analysis](http://usaco.org/current/data/sol_prob3_gold_feb22.html)
+
+## Solution
+
+The subtask of this problem can be solved using brute force: insert all the edges from $N$ points into an edge list and solve using Kruskal's MST algorithm. Upon closer look at our brute force solution, the actual bottleneck is the time to insert all the edges.
+So our focus should be on finding a way to exploit the constraint $0 \le y_i \le 10$, so that the edges we need to insert can be reduced.
+ 
+Given three points $p_1$, $p_2$, and $p_3$, with coordinates $(x_1, y_1)$, $(x_2, y_2)$, and $(x_3, y_3)$ respectively, where $x_1 < x_2 < x_3$. We observe that if $y_1 = y_2$, the length of $p_1 p_3$ will always be larger than $p_1 p_2$ and $p_2 p_3$ individually. According to the nature of Kruskal's MST algorithm, edges being considered are sorted from shortest to longest, and the two pending points have different connected components. The $p_1 p_3$ edge being the longest edge will always be behind $p_1 p_2$ and $p_2 p_3$, and when its turn comes, $p_1$ and $p_3$ are already connected. So we can see that $p_1 p_3$ is an extraneous edge that we can safely remove.
```

Comment:
> and the two pending points have different connected components

how does this relate to kruskal's?

Reply:
Kruskal MST uses the disjoint set union data structure. Points are connected in terms of connected components.

Reply:
hm
how about
> If we follow Kruskal's algorithm, we consider the edges starting from the shortest one, and only add an edge to the tree if it connects two different compoenents.

Reply:
bump

Reply:
bump